File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/CollectorCookie.java
package com.snowplowanalytics.snowplow.network;

import org.json.JSONException;
import org.json.JSONObject;

import okhttp3.Cookie;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;

public class CollectorCookie {

    private final Cookie cookie;

    static List<CollectorCookie> decorateAll(Collection<Cookie> cookies) {
        List<CollectorCookie> collectorCookies = new ArrayList<>(cookies.size());
        for (Cookie cookie : cookies) {
            collectorCookies.add(new CollectorCookie(cookie));
        }
        return collectorCookies;
    }

    CollectorCookie(Cookie cookie) {
        this.cookie = cookie;
    }

    CollectorCookie(String serialized) throws JSONException {
        JSONObject object = new JSONObject(serialized);
        cookie = new Cookie.Builder()
                .name(object.getString("name"))
                .value(object.getString("value"))
                .expiresAt(object.getLong("expiresAt"))
                .domain(object.getString("domain"))
                .path(object.getString("path"))
                .build();
    }

    public boolean isExpired() {
        return cookie.expiresAt() < System.currentTimeMillis();
    }

    Cookie getCookie() {
        return cookie;
    }

    String getCookieKey() {
        return (cookie.secure() ? "https" : "http") + "://" + cookie.domain() + cookie.path() + "|" + cookie.name();
    }

    public String serialize() {
        HashMap<String, Object> values = new HashMap<String, Object>();
        values.put("name", cookie.name());
        values.put("value", cookie.value());
        values.put("expiresAt", cookie.expiresAt());
        values.put("domain", cookie.domain());
        values.put("path", cookie.path());
        return new JSONObject(values).toString();
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof CollectorCookie)) return false;
        CollectorCookie that = (CollectorCookie) other;
        return that.cookie.name().equals(this.cookie.name())
                && that.cookie.domain().equals(this.cookie.domain())
                && that.cookie.path().equals(this.cookie.path());
    }

    @Override
    public int hashCode() {
        int hash = 17;
        hash = 31 * hash + cookie.name().hashCode();
        hash = 31 * hash + cookie.domain().hashCode();
        hash = 31 * hash + cookie.path().hashCode();
        return hash;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/CollectorCookie.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.network

import okhttp3.Cookie
import okhttp3.Cookie.Builder
import org.json.JSONObject

class CollectorCookie {
    val cookie: Cookie

    constructor(cookie: Cookie) {
        this.cookie = cookie
    }

    constructor(serialized: String) {
        val `object` = JSONObject(serialized)
        cookie = Builder()
            .name(`object`.getString("name"))
            .value(`object`.getString("value"))
            .expiresAt(`object`.getLong("expiresAt"))
            .domain(`object`.getString("domain"))
            .path(`object`.getString("path"))
            .build()
    }

    val isExpired: Boolean
        get() = cookie.expiresAt < System.currentTimeMillis()
    
    val cookieKey: String
        get() = (if (cookie.secure) "https" else "http") + "://" + cookie.domain + cookie.path + "|" + cookie.name

    fun serialize(): String {
        val values = HashMap<String?, Any?>()
        values["name"] = cookie.name
        values["value"] = cookie.value
        values["expiresAt"] = cookie.expiresAt
        values["domain"] = cookie.domain
        values["path"] = cookie.path
        return JSONObject(values).toString()
    }

    override fun equals(other: Any?): Boolean {
        if (other !is CollectorCookie) return false

        return other.cookie.name == cookie.name 
                && other.cookie.domain == cookie.domain 
                && other.cookie.path == cookie.path
    }

    override fun hashCode(): Int {
        var hash = 17
        hash = 31 * hash + cookie.name.hashCode()
        hash = 31 * hash + cookie.domain.hashCode()
        hash = 31 * hash + cookie.path.hashCode()
        return hash
    }

    companion object {
        fun decorateAll(cookies: Collection<Cookie>): List<CollectorCookie> {
            val collectorCookies: MutableList<CollectorCookie> = ArrayList(cookies.size)
            for (cookie in cookies) {
                collectorCookies.add(CollectorCookie(cookie))
            }
            return collectorCookies
        }
    }
}
